crates/sui-types/src/messages.rs

```diff
@@ -1328,6 +1329,47 @@ impl TransactionData {
         Self::new_programmable(sender, vec![gas_payment], pt, gas_budget, gas_price)
     }
 
+    pub fn new_upgrade(
+        sender: SuiAddress,
+        gas_payment: ObjectRef,
+        package_id: ObjectID,
+        modules: Vec<Vec<u8>>,
+        dep_ids: Vec<ObjectID>,
+        upgrade_capability: ObjectRef,
+        upgrade_policy: u8,
+        digest: Vec<u8>,
+        gas_budget: u64,
+        gas_price: u64,
+    ) -> Self {
+        let pt = {
+            let mut builder = ProgrammableTransactionBuilder::new();
+            builder
+                .obj(ObjectArg::ImmOrOwnedObject(upgrade_capability))
```

Comment:
There's technically nothing stopping the `upgrade_capability` being a shared object (it's seems an inadvisable thing to do, but it for completeness, it should be possible). cc @patrickkuo on how to make sure we resolve the owner of this object correctly.

Reply:
I actually already have access to the ownership data higher up when I'm grabbing the data for the upgrade capability, so I can pretty easily plumb that information down into there and then we can base the `ObjectArg` here based on that. The update makes sense, so I'll make it so we can take a look at it.